packages/tailwindcss/src/compat/plugin-api.test.ts

```diff
@@ -67,6 +67,18 @@ describe('theme', async () => {
           }
         }
       }
+      @keyframes enter {
+        from {
+          opacity: var(--tw-enter-opacity, 1);
+          transform: translate3d(var(--tw-enter-translate-x, 0), var(--tw-enter-translate-y, 0), 0) scale3d(var(--tw-enter-scale, 1), var(--tw-enter-scale, 1), var(--tw-enter-scale, 1)) rotate(var(--tw-enter-rotate, 0));
+        }
+      }
+      @keyframes exit {
+        to {
+          opacity: var(--tw-exit-opacity, 1);
+          transform: translate3d(var(--tw-exit-translate-x, 0), var(--tw-exit-translate-y, 0), 0) scale3d(var(--tw-exit-scale, 1), var(--tw-exit-scale, 1), var(--tw-exit-scale, 1)) rotate(var(--tw-exit-rotate, 0));
+        }
+      }
```

Comment:
Uh oh wait what.

Reply:
This is interesting, in v3, plugins require an `addBase` to add the keyframes but if it was part of the JS config theme, it would be added automatically. That's lovely:
https://play.tailwindcss.com/YQLQchsWWW?file=config

Reply:
In v3 the way this worked was keyframes were not emitted until used. So you had to do `@keyframes …`: theme('keyframes.whatever`) in your utility function. And we handle that already.

Reply:
Or we do in addUtilities at least. I don't think we do in matchUtilities which we should I think.

Reply:
Or at least I'm like 95% sure it works that way. Maybe we just de-duped them…

Reply:
Oh gotch you're right they're only added when used! If you use `addBase` _and_ use it with an `animate-*` variable, it is emitted twice in v3 (at the end of the stylesheet).

<img width="1291" alt="Screenshot 2024-10-04 at 12 56 42" src="https://github.com/user-attachments/assets/f321617d-1db9-4367-9df4-aee00d7be478">

I guess it's fine if we always emit keyframes for compat now (we do this for v4 keyframes too!). However unsure if we should then guard against this case. Maybe it's ok if the keyframes are emitted twice here since it does no harm? WDYT?

Reply:
yeah it's probably fine — would be nice to dedupe them but that can be an optimization done later. They should still be de-duped when minifying with lightingcss.